map.js

```diff
@@ -195,143 +183,97 @@ function drawMapOverlay(){
-
-            ctx.beginPath();
-            ctx.fillStyle = "black";
-            ctx.textBaseline = "middle";
-            ctx.textAlign = "left";
-            ctx.font = "16px Monospace";
-            ctx.fillText(this.cell.name, this.maxX-mapOffset.x, this.minY + this.height() / 2 - mapOffset.y);
-            ctx.fill();
-        }
-    }
-    ctx.drawImage(this.icon, this.minX-mapOffset.x, this.minY-mapOffset.y, this.width(), this.height());
-    //if(discoveredArr.includes(locObj.formid)){
-    //    ctx.drawImage(icons.Check, this.minX, this.minY, this.width(), this.height());
-    //}
-}
-
 /*********************************
  * TOPBAR FUNCTIONS
  *  this is the "topbar" on the map canvas.
  *********************************/
 
 function initTopbar(){
-    function MapButton(x,y,width,height,text){
+    function MapButton(ordinal,y,height,text){
         MapObject.call(this);
-        this.minX = x;
+        this.ordinal = ordinal;
+        //this.minX and this.maxX calculated by recalculateBoundingBox
         this.minY = y;
-        this.maxX = x+width;
         this.maxY = y+height;
         this.name = text;
-        this.draw = function(ctx){
-            let width = this.width();
-            let height = this.height();
-            ctx.beginPath();
-            if(currentOverlay == this.name){
-                ctx.fillStyle = "#ccc";
-            }
-            else{
-                ctx.fillStyle = "#E5D9B9";
-            }
-            ctx.fillRect(this.minX, this.minY, width, height);
-
-            //and now text
-            ctx.beginPath();
-            ctx.fillStyle = "#000000";
-            ctx.textAlign = "center";
-            ctx.textBaseline = "alphabetic";
-            ctx.font = "16px Arial"
-            ctx.fillText(this.name, this.minX + (width/2), this.minY + 16);
-        }
+        this.recalculateBoundingBox();
+        
     }
     MapButton.prototype = Object.create(MapObject.prototype);
+    MapButton.prototype.draw = function(ctx){
+        let width = this.width();
+        let height = this.height();
+        ctx.beginPath();
+        if(currentOverlay == this.name){
+            ctx.fillStyle = "#ccc";
+        }
+        else{
+            ctx.fillStyle = "#E5D9B9";
+        }
+        ctx.fillRect(this.minX, this.minY, width, height);
 
-    topbar = new MapObject();
-    topbar.buttons = [];
-    topbar.minX = 0;
-    topbar.minY = 0;
-    topbar.maxX = viewport.clientWidth;
-    topbar.maxY = 32;
+        //and now text
+        ctx.beginPath();
+        ctx.fillStyle = "#000000";
+        ctx.textAlign = "center";
+        ctx.textBaseline = "alphabetic";
+        ctx.font = "16px Arial"
+        ctx.fillText(this.name, this.minX + (width/2), this.minY + 16);
+    }
+    MapButton.prototype.recalculateBoundingBox = function(){
+        this.minX = 8 + (map_topbar.width()/3)*this.ordinal;
+        this.maxX = this.minX + (map_topbar.width() - 16)/3;
+    }
+
+    map_topbar = new MapObject();
```

Comment:
Unlike the overlay, we want to intercept all clicks that happen on the topbar space, even if it's between 2 buttons.
So it is a MapObject itself.
And then, all the child buttons are called from within this object's methods.
(This parent-child model is similar to what many Real UI systems, like java.swing and Windows, do)